Marlin/src/HAL/HAL_ESP32/fastio_ESP32.cpp

```diff
@@ -0,0 +1,62 @@
+/**
+ * Marlin 3D Printer Firmware
+ * Copyright (C) 2016 MarlinFirmware [https://github.com/MarlinFirmware/Marlin]
+ *
+ * Based on Sprinter and grbl.
+ * Copyright (C) 2011 Camiel Gubbels / Erik van der Zalm
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <Wire.h>
+#include "fastio_ESP32.h"
+
+uint8_t get_i2c_register(uint8_t reg) {
+  Wire.beginTransmission(GPIOX_I2C_ADDR);
+  Wire.write(reg);
+  Wire.endTransmission();
+
+  Wire.requestFrom(GPIOX_I2C_ADDR, 1);
+  return Wire.read();
+}
+
+void set_i2c_register(uint8_t reg, uint8_t val) {
+  Wire.beginTransmission(GPIOX_I2C_ADDR);
+  Wire.write(reg);
+  Wire.write(val);
+  Wire.endTransmission();
+}
+
+void set_i2c_bit(uint8_t reg, uint8_t bit, uint8_t val) {
+  uint8_t r = get_i2c_register(reg);
+  r = val ? (r | (1 << bit)) : (r & ~(1 << bit));
+  set_i2c_register(reg, r);
+}
+
+void set_i2c_pin_mode(uint8_t pin, uint8_t mode) {
+  set_i2c_bit(GET_REGISTER(pin, IODIRA), GET_PIN(pin), mode & INPUT ? 1 : 0);
```

Comment:
Note that in `Arduino.h`, `INPUT` is defined as 0. So `mode & INPUT` will always result in 0. I'm guessing that `INPUT` here should be something else. If I can figure it out I'll make the edit.

Reply:
Hi @thinkyhead, thanks for the feedbacks, I  just double checked and in the ESP32 definition INPUT is 1 (https://github.com/espressif/arduino-esp32/blob/master/cores/esp32/esp32-hal-gpio.h#L33)

Reply:
Is that compatible with the `pinMode(pin, INPUT)` function?

Reply:
I guess you mean the fact that Input is not 0? Yes it does work fine, if you look at the HAL I used pinMode for SET_INPUT/SET_OUTPUT for the esp32 pins.

The SET_INPUT/SET_OUTPUT functions will use `set_i2c_pin_mode` if the pin is on the GPIO expander and use `pinMode(pin, INPUT)` if it's one of the ESP pin.